File: codes/julia/ch20/behavioral_bc_laguerre.jl
# behavioral_bc_laguerre.jl
# Chapter 20: Spectral Methods on Unbounded Intervals
# Computational Etude 20.7: Behavioural boundary condition at infinity.
#
# Author: Dr. Denys Dutykh (Khalifa University, Abu Dhabi, UAE)

using CairoMakie
using LinearAlgebra
using Printf

include(joinpath(@__DIR__, "..", "ch07", "cheb_matrix.jl"))

function tln_dmatrices(N, L)
    Dx, x = cheb_matrix(N)
    fp = 2L ./ (1 .- x) .^ 2
    fpp = 4L ./ (1 .- x) .^ 3
    Dy = Diagonal(1 ./ fp) * Dx
    Dy2 = Diagonal(1 ./ fp .^ 2) * (Dx * Dx) - Diagonal(fpp ./ fp .^ 3) * Dx
    y_full = L .* (1 .+ x) ./ (1 .- x)
    return y_full[2:end], Dy[2:end, 2:end], Dy2[2:end, 2:end]
end

function solve_A(N, L)
    y, Dy, Dy2 = tln_dmatrices(N, L)
    Y = Diagonal(y)
    A = Y * Dy2 + (Y + I) * Dy
    return sort(real.(eigen(-Matrix(A)).values))
end

function solve_B(N, L)
    y, Dy, Dy2 = tln_dmatrices(N, L)
    Y = Diagonal(y); M = Diagonal(0.5 .+ 0.25 .* y)
    A = Y * Dy2 + Dy - M
    return sort(real.(eigen(-Matrix(A)).values))
end

function count_good(eigs; tol=0.05)
    targets = 0:49
    taken = Set{Int}()
    good = 0
    for lam in eigs
        if lam < -0.5
            continue
        end
        order = sortperm(abs.(lam .- targets))
        for idx in order
            n = targets[idx]
            if n in taken
                continue
            end
            rel = abs(lam - n) / max(1.0, n)
            if rel < tol
                push!(taken, n); good += 1
            end
            break
        end
    end
    return good
end

function run()
    outdir = joinpath(@__DIR__, "..", "..", "..",
                      "textbook", "figures", "ch20", "julia")
    mkpath(outdir)
    L = 32.0
    Ns = [10, 20, 30, 40, 60, 80, 120]
    good_A = [count_good(solve_A(N, L)) for N in Ns]
    good_B = [count_good(solve_B(N, L)) for N in Ns]

    NAVY = colorant"#142D6E"; CORAL = colorant"#E74C3C"; TEAL = colorant"#16A085"
    fig = Figure(size=(1020, 340))

    ax1 = Axis(fig[1, 1], xlabel="N", ylabel="good eigenvalues",
               title="(a) Good-eigenvalue count")
    scatterlines!(ax1, Ns, good_A; color=CORAL, label="naive")
    scatterlines!(ax1, Ns, good_B; color=TEAL, label="behavioural")
    axislegend(ax1; position=:lt)

    eigs_A = solve_A(40, L)[1:20]
    eigs_B = solve_B(40, L)[1:20]
    ax2 = Axis(fig[1, 2], xlabel="Re(lambda)", ylabel="Im(lambda)",
               title="(b) Spectrum at N=40",
               limits=((-1, 20), (-3, 3)))
    scatter!(ax2, real.(eigs_A), imag.(eigs_A); color=CORAL, label="A naive")
    scatter!(ax2, real.(eigs_B), imag.(eigs_B); color=TEAL, label="B behavioural", marker=:rect)
    hlines!(ax2, [0]; color=:gray, linewidth=0.5)
    axislegend(ax2; position=:rt)

    save(joinpath(outdir, "behavioral_bc_laguerre.pdf"), fig)
    save(joinpath(outdir, "behavioral_bc_laguerre.png"), fig)
    @printf("[20.7-julia] saved figure\n")
end

if abspath(PROGRAM_FILE) == @__FILE__
    run()
end

File: codes/matlab/ch20/behavioral_bc_laguerre.m
function behavioral_bc_laguerre()
%% behavioral_bc_laguerre - Chapter 20, Etude 20.7.
% Boundedness is sometimes enough, sometimes not.
% Author: Dr. Denys Dutykh.

    script_dir = fileparts(mfilename('fullpath'));
    addpath(fullfile(script_dir, '..', 'ch07'));
    out_dir = fullfile(script_dir, '..', '..', '..', ...
                       'textbook', 'figures', 'ch20', 'matlab');
    if ~exist(out_dir, 'dir'); mkdir(out_dir); end

    ell = 32;
    Ns = [10 20 30 40 60 80 120];
    good_A = zeros(size(Ns)); good_B = zeros(size(Ns));
    for i = 1:length(Ns)
        eA = sort(real(solve_A(Ns(i), ell)));
        eB = sort(real(solve_B(Ns(i), ell)));
        good_A(i) = count_good(eA);
        good_B(i) = count_good(eB);
    end

    NAVY=[20 45 110]/255; CORAL=[231 76 60]/255; TEAL=[22 160 133]/255;
    fig = figure('Position',[100 100 1020 340],'Color','w');
    subplot(1,2,1);
    plot(Ns, good_A, '-o', 'Color', CORAL); hold on;
    plot(Ns, good_B, '-s', 'Color', TEAL);
    xlabel('N'); ylabel('good eigenvalues'); grid on; box on;
    title('(a) Good-eigenvalue count'); legend({'naive','behavioural'});

    subplot(1,2,2);
    eA = solve_A(40, ell); eB = solve_B(40, ell);
    plot(real(eA(1:20)), imag(eA(1:20)), 'o', 'Color', CORAL); hold on;
    plot(real(eB(1:20)), imag(eB(1:20)), 's', 'Color', TEAL);
    yline(0, '-', 'Color', [0.5 0.5 0.5]);
    xlim([-1 20]); ylim([-3 3]); grid on; box on;
    xlabel('Re(\lambda)'); ylabel('Im(\lambda)');
    title('(b) Spectrum (N=40)');
    legend({'A naive','B behavioural'});

    print(fig, fullfile(out_dir, 'behavioral_bc_laguerre.pdf'), '-dpdf');
    print(fig, fullfile(out_dir, 'behavioral_bc_laguerre.png'), '-dpng');
    fprintf('[20.7-matlab] figure saved\n');
end

function [y, Dy, Dy2] = tln_dmatrices(N, ell)
    [Dx, x] = cheb_matrix(N);
    fp = 2*ell ./ (1-x).^2;
    fpp = 4*ell ./ (1-x).^3;
    Dy = diag(1./fp) * Dx;
    Dy2 = diag(1./fp.^2)*(Dx*Dx) - diag(fpp./fp.^3)*Dx;
    y_full = ell*(1+x)./(1-x);
    y = y_full(2:end);
    Dy = Dy(2:end, 2:end);
    Dy2 = Dy2(2:end, 2:end);
end

function e = solve_A(N, ell)
    [y, Dy, Dy2] = tln_dmatrices(N, ell);
    Y = diag(y);
    A = Y*Dy2 + (Y + eye(length(y)))*Dy;
    e = eig(-A);
end

function e = solve_B(N, ell)
    [y, Dy, Dy2] = tln_dmatrices(N, ell);
    Y = diag(y); M = diag(0.5 + 0.25*y);
    A = Y*Dy2 + Dy - M;
    e = eig(-A);
end

function n = count_good(eigs)
    eigs = eigs(eigs > -0.5);
    targets = 0:49;
    taken = false(size(targets));
    n = 0;
    for i = 1:length(eigs)
        [~, order] = sort(abs(eigs(i) - targets));
        for k = 1:length(order)
            idx = order(k);
            if taken(idx), continue; end
            rel = abs(eigs(i) - targets(idx)) / max(1, targets(idx));
            if rel < 0.05
                taken(idx) = true; n = n + 1;
            end
            break;
        end
    end
end
